### PR TITLE
GSoC fixes

### PR DESCRIPTION
**Pull request by garbear**
## Description

This PR contains fixes from the last four years of GSoC projects. Includes:

* One fix from GSoC 2017 - https://github.com/garbear/xbmc/pull/86
* Code path optimization from GSoC 2020 - https://github.com/garbear/xbmc/pull/116
* Controller fix from GSoC 2020 - https://github.com/garbear/xbmc/pull/119
* Three random fixes I made over the years

## Motivation and context

Hoping to land one or two GSoC PRs to master this summer. I've been building these fixes in my `retroplayer` branches and figured now was a good time to upstream them.

## How has this been tested?

Compile-tested on Win64. Runtime tested the fix on Linux.

## What is the effect on users?

The only visible effect is the fix:

* Fixed similar controllers interfering with each other while button mapping

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

File: tools/windows/prepare-binary-addons-dev.bat [hunk 1]
```diff
@@ -76,7 +76,7 @@ IF "%addon%" NEQ "" (
 )
 
 rem execute cmake to generate Visual Studio 12 project files
```
Comment (fuzzard) [inline tools/windows/prepare-binary-addons-dev.bat:78]: i dont want to get into how windows uses this, but maybe update this comment from "Visual Studio 12" just to be inline with the actual cmake generator.

My query would be, should the generator be hard coded, or should an env var or whatever be used, with a potential fall back to a "default" generator version if not set?

File: xbmc/guilib/GUIControlFactory.cpp [hunk 2]
```diff
@@ -271,12 +272,14 @@ bool CGUIControlFactory::GetDimensions(const TiXmlNode *node, const char *leftTa
...
         left = center - width/2;
-        hasLeft = hasWidth = true;
+        hasLeft = true;
+        hasWidth = true;
       }
     }
     else if (hasLeft) // neither right nor center specified
     {
       width = std::max(0.0f, parentSize - left); // if left=0, this fills to parent
+      hasWidth = true;
```
Comment (fuzzard) [inline xbmc/guilib/GUIControlFactory.cpp:282]: Maybe im not following this completely, but these changes seem to just set Haswidth/hasleft true for every codepath. Therefore the return will always just be true?

Couldnt just the return be changed to true, and leave everything as is?

Comment (fuzzard) [reply xbmc/guilib/GUIControlFactory.cpp:282]: To add to that, wouldnt this change effectively make this block obsolete?

https://github.com/xbmc/xbmc/blob/8ababf52c9604419d12dad8f214ec2b8f578be3e/xbmc/guilib/GUIControlFactory.cpp#L785-L802

I havent looked into the history of any of that, so dont have any further context around its use.

File: tools/windows/prepare-binary-addons-dev.bat (hunk 1 above)
Comment (garbear) [reply tools/windows/prepare-binary-addons-dev.bat:78]: This is just a development script (note the path, production scripts use `tools/buildsteps/windows`) so hard-coding is fine.

**Review by VelocityRa — COMMENTED** [review]
huh thought i had pushed this

File: xbmc/guilib/GUIControlFactory.cpp (hunk 2 above)
Comment (garbear) [reply xbmc/guilib/GUIControlFactory.cpp:282]: > Couldnt just the return be changed to true, and leave everything as is?

From a propositional standpoint, there's no fall-through else, so valid code paths can have false be returned. I don't see how to make anything obsolete without changing the logic.

**Review by VelocityRa — APPROVED** [review]

**Pull request merged**